src/main/java/org/jboss/remoting3/remote/OutboundMessage.java

```diff
@@ -104,36 +104,30 @@ public void accept(final Pooled<ByteBuffer> pooledBuffer, final boolean eof) thr
-                        if (window >= msgSize) {
-                            window -= msgSize;
-                            if (log.isTraceEnabled()) {
-                                log.tracef("Outbound message ID %04x: message window is open (%d-%d=%d remaining), proceeding with send", getActualId(), window + msgSize, msgSize, window);
-                            }
-                            break;
-                        }
-                        try {
-                            log.tracef("Outbound message ID %04x: message window is closed, waiting", getActualId());
-                            pipeOutputStream.wait(ackTimeout, 0);
-                            if (window == currentWindow) {
-                                // no changes, throw an exception
-                                timeoutExpired = true;
+                    if (!decrementWindow(msgSize)) {
+                        long ackTimeout = OutboundMessage.this.ackTimeout;
+                        long initialTime = System.currentTimeMillis();
+                        do {
+                            try {
+                                log.tracef("Outbound message ID %04x: message window is closed, waiting", getActualId());
+                                pipeOutputStream.wait(ackTimeout, 0);
+                            } catch (InterruptedException e) {
+                                cancelled = true;
+                                intr = true;
                                 break;
                             }
-                        } catch (InterruptedException e) {
-                            cancelled = true;
-                            intr = true;
-                            break;
-                        }
-                        if (closeReceived) {
-                            throw new BrokenPipeException(this + ": remote side closed the message stream");
-                        }
-                        if (closeCalled && ! eof) {
-                            throw new NotOpenException(this + ": message was closed asynchronously by another thread");
-                        }
-                        if (cancelSent) {
-                            throw new MessageCancelledException(this + ": message was cancelled");
+                            if (closeReceived) {
+                                throw new BrokenPipeException(this + ": remote side closed the message stream");
+                            }
+                            if (closeCalled && ! eof) {
+                                throw new NotOpenException(this + ": message was closed asynchronously by another thread");
+                            }
+                            if (cancelSent) {
+                                throw new MessageCancelledException(this + ": message was cancelled");
+                            }
+                        } while (!decrementWindow(msgSize) && (ackTimeout -= (System.currentTimeMillis() - initialTime)) > 0);
```

Comment:
Is it intentional that `ackTimeout` is always reduced by the total time elapsed, rather than the time since the last loop?

Reply:
Aha, now I see why @ropalka  found it strange that I was initiating the initialTime before the loop. Thank you both for catching this. This is not correct, I'll fix it.